tests/testthat/test-GRN.R

```diff
@@ -0,0 +1,44 @@
+# Test GRN function
+
+# Run example of GRN
+data('DEGsmatrix')
+data('dataFilt')
+dataDEGs <- DEGsmatrix
+set.seed(317)
+dataGRN_test <- GRN(TFs = sample(rownames(dataDEGs), 100),
+                    DEGsmatrix = dataDEGs,
+                    DiffGenes = TRUE,
+                    normCounts = dataFilt,
+                    nGenesPerm = 5,
+                    nBoot = 5)
+
+# Load example data of GRN serving as reference point
+data(dataGRN)
+
+# Test that output of GRN is as expected
+
+# Test that output of GRN is a list of length 2
+test_that("output of GRN is a list", {
+  expect_type(dataGRN_test, "list")
+  expect_type(dataGRN_test$miTFGenes, "double")
+  expect_type(dataGRN_test$maxmi, "double")
+  expect_length(dataGRN_test, 2)
+})
+
+# Test that names of elements in GRN are correct
+test_that("names of elements in GRN output are correct", {
+  expect_equal(names(dataGRN_test), c("miTFGenes", "maxmi"))
+})
+
+# Test that output of GRN is as expected compared to reference / example data
+test_that("GRN output is identical to reference point", {
+  expect_equal(dataGRN_test, dataGRN, tolerance = 0.1)
```

Comment:
0.1 is a bit high as tolerance for MI; ideally we should define tolerance depending on the significant digits of the output data type. Can we do at least third decimal?

Reply:
It is not possible because that will lead to failure of the test because sometimes the MI can be e.g. 0.29 and then the next time it can be 0.30, hence the tolerance = 0.1. I believe there is a stochastic nature of the parmigene package that occurs even though we set a seed; in R, you need to set the seed right before you run a line so I don't think this seed is registered within the parmigene functions but at least by setting the seed like this we are getting consistent results with tolerance = 0.1. 

If you have other ideas let me know

Reply:
I don't think that's how `expect_equal` works, it checks if the difference is at least of 0.1, e.g.

```
> expect_equal(0.2998, 0.3, tolerance=0.1)
>

> expect_equal(0.2998, 0.3999, tolerance=0.1)
Error: 0.2998 not equal to 0.3999.
1/1 mismatches
[1] 0.3 - 0.4 == -0.1
```

if the difference between two runs is as large as 0.1, I think our test is not going to be very accurate. What we would need to do is to diminish the uncertainty of the result here rather than have very large (relatively speaking) allowed tolerance. I tried having a look at the source code of `knnmi.cross`, it seems to me that it is non-deterministic in generating a random seed that is used to generate noise. The order of magnitude of noise can be set in the `knnmi.cross` function call, I did a few tests and I think it becomes completely reproducible if we set it to `0`, e.g.:

```
m1 = random_matrix <- matrix(runif(n=10000, min=1, max=20), nrow=1000, ncol=100)
m2 = random_matrix <- matrix(runif(n=10000, min=1, max=20), nrow=1000, ncol=100)

d = knnmi.cross(m1, m2, k=2, noise=0)

print(d[10,10])

d = knnmi.cross(m1, m2, k=2, noise=0)

print(d[10,10])
```

I always get the same result:

```
(py37) teo@kb-bioinfo01:teo$ Rscript test.R
[1] 0.2416632
[1] 0.2416632
(py37) teo@kb-bioinfo01:teo$ Rscript test.R
[1] 0.2416632
[1] 0.2416632
(py37) teo@kb-bioinfo01:teo$ Rscript test.R
[1] 0.2416632
[1] 0.2416632
(py37) teo@kb-bioinfo01:teo$ Rscript test.R
[1] 0.2416632
[1] 0.2416632
(py37) teo@kb-bioinfo01:teo$ Rscript test.R
[1] 0.2416632
[1] 0.2416632
```

worth a try?

Reply:
we could e.g. add a new small dataset with these "reproducible" results specifically for this test, not sure it is worth resaving it as the new GRN dataset and then redoing all the other results/tests in the pipeline, if they are reproducible